### ISSUE TITLE
Puma server is hung! URGENT

### ISSUE THREAD

**Issue by BlackNightFury** [puma/puma#1608]
We are running Puma with Sinatra
We are running 2 workers with 2 CPUs. And when we try concurrent request, workers use 100% cpu and it is remaining around 100% even though we finish concurrent request. And it is hanging there.

Puma version : 3.11.4
This is my puma configuration. 

```
require 'puma'
require 'inifile'

root = "#{Dir.getwd}"
#Server Configuration Filename
serverConf_file = "#{root}/server.ini"

#check if server configuration file exists
if File.exist?(serverConf_file)
  puts "Reading server configuration from #{serverConf_file}"

  #load server configuration file
  myini = IniFile.load(serverConf_file)

  #loop all sections
  myini.each_section do |section|

    # save defulat server setting
    if(section == "Default")

      myini[section].each { |key, value|
        if key == "IPAddress"
          $localhostName = value
        elsif key == "Port"
          $localhostPort = value
        elsif key == "Use_TLS"
          $use_tls = value
        end
      }
    end
  end
else
  puts "No configuration file!"
  exit(1)
end

$global_flag = 'Global'
puts   'Reading puma configuration'

if $use_tls
  if $localhostPort == 443
    $baseUrl = "https://#{$localhostName}"
  else
    $baseUrl = "https://#{$localhostName}:#{$localhostPort}"
  end
else
  if $localhostPort == 80
    $baseUrl = "http://#{$localhostName}"
  else
    $baseUrl = "http://#{$localhostName}:#{$localhostPort}"
  end
end

puts "Binding to #{$baseUrl}"

if $use_tls
  bind "ssl://#{$localhostName}:#{$localhostPort}?key=#{root}/server.key&cert=#{root}/server.crt&verify_mode=none&ca=#{root}/intermediate.crt"
else
  bind "tcp://#{$localhostName}:#{$localhostPort}"
end
pidfile "#{root}/tmp/pids/puma.pid"
rackup "#{root}/config.ru"
state_path "#{root}/tmp/pids/puma.state"

#stdout_redirect "#{root}/log/stdout", "#{root}/log/stderr"
worker_timeout 300

workers 2
threads 5, 15
daemonize true
```
This is the Sinatra configuration

```
configure do
  puts "settings before: #{settings}"
#  hides stack trace
# Show errors to client aps (disable for prod)
  set :show_exceptions, true

#  messes up erb template
#  mime_type :plain, 'text/ plain'

# one of production, test, or development (standard vars), custom symbols
#  like 'conformance' allowed too
   set :root, File.dirname(__FILE__)
   set :server, :puma
# don't serve static files even in public_folder exists
  set :static, false

# ignored for now, just want to be explicit
  set :public_folder, 'c:\sinatra\\'

# display errors to admin console
  set :dump_errors, true

  enable :cross_origin
end
```

Puma server is hung when I try to do concurrent request. It is happening very randomly.

When I try to stack tract it with gdb, it is showing this:

```
from /usr/local/rvm/gems/ruby-2.2.6/gems/puma-3.11.4/lib/puma/thread_pool.rb:244:in `block in start!'
        from /usr/local/rvm/gems/ruby-2.2.6/gems/puma-3.11.4/lib/puma/thread_pool.rb:244:in `sleep'

```

I ran the puma server by `bundle exec puma`
This is production usage. So please help me to fix this issue asap. Then it will be appreciated.
Thanks

**Comment by benoittgt** [comment]
Please @BlackNightFury this is an open source project. No one is pay to maintain this. 
> So please help me to fix this issue asap

I don't think "asap" is needed if your situation is critical. :)

**Comment by BlackNightFury** [comment]
@benoittgt Thanks for your attention. But this project is very confidential and we don't use even github.
I already shared puma configuration. It works with Passenger with no problem.

But it is hanging with Puma. I think this is because of minissl problem.

But don't know how to figure it out. It is working with http puma fine but not working over ssl.

**Comment by timkrins** [comment]
I noticed a SSL fix in the latest release (3.12.0) - maybe give that a try
